### PR TITLE
Runtime code refactoring: `domain_is_foo`

### PR DESCRIPTION
**Pull request by gasche**
TL;DR: This is a refactoring PR that was prompted by a comment of @kayceesrk when reviewing #12597.

In #12597 I needed a version of `caml_domain_is_terminating(void)` that takes a `caml_domain_state *` as parameter, because I need to ask "are we in the process of terminating?" within a function that is itself parametrized on the domain state. (In fact that function is only ever called with the current domain, but that is another story.)
I considered simply changing `caml_domain_is_terminating(void)` to take a `caml_domain_state *` parameter, but there are other `caml_domain_is_foo` functions and they all take no arguments, so this would have been inconsistent. I chose instead to introduce another function `caml_domain_terminating(caml_domain_state *)` in addition to the existing one.

@kayceesrk suggested in review to push back and move to the `caml_domain_is_terminating(caml_domain_state *)` version. In the present PR, I do this, but I also change the other `caml_domain_is_foo` functions for consistency. The result is a non-trivial refactoring PR:

1. `caml_domain_is_in_stw` now takes a `caml_domain_state *` parameter, this was a trivial change
2. `caml_domain_is_terminating` now takes a `caml_domain_state *` parameter, this is a simple change which is needed for #12597 anyway
3. The hard one is `caml_domain_is_multicore(void)`. It does not really make sense to make it a domain-parametrized function, as this is a global property of the runtime, not of a single given domain. (But then currently the code does not work if it is not called from an OCaml domain thread). So I renamed it to `caml_runtime_is_multicore` *and* changed the implementation to be simpler, and safe to call from any domain. (We just have a global atomic boolean in the runtime to remember if `spawn` was ever called.)
4. I left `caml_domain_alone()`... alone for now because it is used in several places in the standard library, making any change there more invasive (I think it could be called `caml_runtime_is_single_domain` instead).

I am left wondering whether the change is worth it. The result is certainly more consistent / a slightly nicer API, but this is some amount of small code churn, and the change of implementation of `caml_domain_is_multicore` has the potential for hidden surprises.

I would propose to let @kayceesrk judge this, if he has the time to look at the PR.

Note: `caml_domain_is_multicore` returns an `intnat`, while `caml_domain_is_in_stw` and `caml_domain_is_terminating` returns an `int`. I don't know if the inconsistency is intentional or not, and left it alone with no change. (I wish we had a boolean type defined in the runtime.)

### CODE REVIEW

**Comment by gasche** [comment]
I am not convinced that this PR is worth the trouble: to get a consistent naming convention, it changes the behavior of some of the affected functions in subtle ways, and I think that the cost of reviewing those changes probably outweigh the consistency gains. Closing.
